### PR TITLE
Add remote borg backup support

### PR DESCRIPTION
**Pull request by timdiels**
This adds support for backing up directly to a remote borg repo. ([I tried reaching out earlier](https://github.com/nextcloud/all-in-one/discussions/1068#discussioncomment-8907198))

### Why?
I'd like to backup/restore via the UI, resist ransomware and store the backups remotely without storing a local (compressed) copy of the nextcloud data.

Alternatives (https://github.com/nextcloud/all-in-one?tab=readme-ov-file#are-remote-borg-backups-supported):

- Mount a network file system: Ransomware can encrypt the backups on the network storage.
- rclone: Ransomware and stores a local copy of the borg repo
- [borg backup the local repo to borgbase](https://github.com/nextcloud/all-in-one/discussions/4391): Stores a local compressed copy of the nextcloud data as at least 1 borg backup
- create your own solution: backups cannot be managed through the UI and others will have to reinvent my work

### How to use it
Instead of entering a local repo path (which remains supported), you can choose to instead enter a remote borg repo url. (These screenshots are from a disaster recovery but it's analogous, just ignore the passphrase):

Screenshot from 2024-06-09 15-26-41

The first you try to initialise the repo, the backup container will create an ssh key and foolishly try to init the borg repo resulting in error because you have to authorise the ssh key it generated first. So, e.g. I copy paste the public key shown here to my borgbase repo:

Screenshot from 2024-06-09 15-27-01

Then I can try again as instructed and it should work. The following backup info is shown:

Screenshot from 2024-06-09 14-59-30

### Changes
- Support for storing backups directly at a remote borg repo (Replace commands to not assume the repo is local, e.g. use BORG_REPO env var, ...)
- Exclude files from backups if we don't plan on restoring them anyway, such as the audit log
- Restore with `borg extract` because, at least with a remote repo, `borg mount` is very slow (20 seconds vs 24 minutes restore for pretty much an empty nextcloud install).

FYI Oddly the original code never restores host-mounts.

### Testing
I hacked the code til I could run local deployments of nextcloud-aio without any builds and manually tested:

- Entering invalid stuff into forms I changed
- Remote borg repo init
- Creating and restoring a backup. I created some files in the master container volume; some in an ignored directory to make sure those are not deleted, some in a directory that is part of the backup, after the backup made some changes and made sure that once restored, the ignored files were left alone, new (non-ignored) files were deleted and changed files restored. I can't really open the user UI with my local test setup.
- Checking a backup
- Setting up a new nextcloud from backup

### TODO
- I did at one point test local backups, but I should redo that.
- If we could do a dev build that I could tmp upgrade my actual instance to for testing real backups, that would be great
- Update the readme, but I first want to hear your thoughts before putting more time in it

### CODE REVIEW

**Comment by szaimen** [comment]
Hi, first of all thank you for your contribution!

I've thought a bit about this the last week and I still do not want to maintain this since it requires an additional server for me for testing and has the potential of many users needing help getting this to work. However I would be fine with merging this if you @timdiels would step in as the maintainer of this feature. That means if bug reports or any questions regarding this feature come in, I would ask you for help on the topic. If that is fine for you, I would continue with the review.

> FYI Oddly the original code never restores host-mounts.

Yes, this is expected and documented.

> ### TODO
> * I did at one point test local backups, but I should redo that.

Sounds good!

> * If we could do a dev build that I could tmp upgrade my actual instance to for testing real backups, that would be great

Yeah, we could create a dev instance from this if we proceed...

> * Update the readme, but I first want to hear your thoughts before putting more time in it

Yes

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [inline Containers/borgbackup/backupscript.sh:334]: I would personally like to keep rsync at least for local archives since as you found out borg extract does not allow for deleting existing files...

Comment (timdiels) [reply Containers/borgbackup/backupscript.sh:334]: Sure, I'll bring back the borg mount and rsync for local repos.

**Comment by timdiels** [comment]
> However I would be fine with merging this if you @timdiels would step in as the maintainer of this feature. That means if bug reports or any questions regarding this feature come in, I would ask you for help on the topic. If that is fine for you, I would continue with the review.

I'm willing to maintain it, mostly on weekends, for as long as I'm a nextcloud user (which I have been for a couple of years so far).

> > * I did at one point test local backups, but I should redo that.

I will do it on the dev build after I've made the change you suggested.

> > * If we could do a dev build that I could tmp upgrade my actual instance to for testing real backups, that would be great

Is there an easier way to develop than the hacks I did (in a different branch) for local testing? Not sure how you normally test AIO? If it's always via push to github, wait for build and deploy, I will probably keep my debug branch for later.

Note to self:

- [ ] Update readme
- [ ] Local repo should use borg mount and rsync

**Comment by szaimen** [comment]
> I'm willing to maintain it, mostly on weekends, for as long as I'm a nextcloud user (which I have been for a couple of years so far).

Cool, then we can go ahead with this PR. I've invited you to the repo for easier collaboration :) 

> > > * I did at one point test local backups, but I should redo that.
> 
> I will do it on the dev build after I've made the change you suggested.

Great :)

> > > * If we could do a dev build that I could tmp upgrade my actual instance to for testing real backups, that would be great
> 
> Is there an easier way to develop than the hacks I did (in a different branch) for local testing? Not sure how you normally test AIO? If it's always via push to github, wait for build and deploy, I will probably keep my debug branch for later.

Usually it is indeed always via push to github, wait for build and deploy. So probably it is easier for you to keep your debug branch 👍

**Comment by gregjohnsonsaltaire** [comment]
Very much appreciate the PR guys!
I'm happy to help test informally ... I have 3 nextcloud-aio sites of 20GB, 40GB & 350GB doing borg backups to Hetzner Storageboxes that I'm prepared to convert to the PR ... Regards

File: Containers/borgbackup/backupscript.sh
Comment (asavageiv) [inline Containers/borgbackup/backupscript.sh:140]: Should this be "Likely `borg init` will error on a partially initialized repo"?

Comment (timdiels) [reply Containers/borgbackup/backupscript.sh:140]: Probably most borg commands would fail but it's referring to the borg info on line 120. If we ever end up with a partially initialized repo (sounds very rare), I think manual intervention is required anyway? The important thing is I believe `borg info` will fail on such a corrupt repo and will end up at line 124 instead of trying to initialize it again. The config file was used as a sentinel to detect whether the repo has been initialized already (see the deleted line 107), which why it is removed here and why it's probably no longer necessary to do this.

**Comment by timdiels** [comment]
I tried out local backups on my machine and made all the requested changes, I believe. Could we get a a dev build going?

File: Containers/borgbackup/backupscript.sh
Comment (timdiels) [reply Containers/borgbackup/backupscript.sh:334]: Done

File: Containers/mastercontainer/Dockerfile
Comment (szaimen) [inline Containers/mastercontainer/Dockerfile:69]: TODO: revert this later on

**Review by szaimen — CHANGES_REQUESTED** [review]

**Comment by szaimen** [comment]
> I tried out local backups on my machine and made all the requested changes I believe. 

Nice!

> Could we get a a dev build going?

Yes, it should be available via nextcloud/all-in-one:develop-4804

**Comment by timdiels** [comment]
I'm trying it out on my own server (with a VM snapshot before switching to dev). Will I be able to later update this to `:latest` after it's included in a release or should I revert my snapshot after testing instead?

**Comment by timdiels** [comment]
Created a backup on my server, I'll enable autobackup/update and try out restoring after some changes later (might be next weekend):

```
Duration: 1 hours 2 minutes 27.77 seconds
Number of files: 34348
Utilization of max. archive size: 0%
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:                9.93 GB              8.89 GB              8.65 GB
```

**Comment by timdiels** [comment]
I rolled back to old snapshot because apparently the dev build was older than the latest released nextcloud, so it wouldn't start anymore. I'll try again on the weekend with a tmp vm instead.

**Comment by msullivan-lc** [comment]
Appreciate all the work you are doing on this @timdiels. Very much looking forward to this addition!

**Comment by devnoname120** [comment]
@szaimen Is something missing before this pull request can be merged? I see it keeps being pushed back to the next release. I'm not sure of the significance of this as I'm not very familiar with the release process that Nextcloud AIO follows. Thanks!

**Comment by szaimen** [comment]
Mostly I am waiting for the results of the tests that @timdiels was doing...

Afterwards this needs conflict resolution, review and testing...

**Comment by timdiels** [comment]
Is there a way to get it to rebuild the develop-4804 image? My last change is not included in the image.

**Comment by timdiels** [comment]
Retested remote and local backups with the dev image and it all seems to work. This time I added a local caddy so I could actually test with adding a user, some files, some notes through the UI.

**Comment by timdiels** [comment]
Up to date again, I mostly only saw a switch from `<br>` to `<p>` in the merge conflict.

There's the one change in the backupscript that I haven't been able to test because I need a rebuild of the dev image for that. Other than that it's a go for me.

**Comment by msullivan-lc** [comment]
Hi @szaimen, my turn for a gentle ask on the status of this one? @timdiels seems to have brought things up to date and is ready to test. Really looking forward to this feature so I can roll out Nextcloud for my organization. Easy backup to a remote location is the only thing holding me up. I've tried several of the other documented methods but found them to either be unreliable or too complicated.

Please let us know. 

Thanks!

**Comment by szaimen** [comment]
Hi @timdiels, sorry for my late reply. I was on vacation.

So my plan would be to include this for the next major release of AIO which will likely be released after Nc30.0.1 or 30.0.2 is out. Until then we have a bit of time to finish this.

I for now rebased your branch and pushed my attempts to https://github.com/nextcloud/all-in-one/pull/5206. If possible, I would like to continue working on that branch. Please review my rebase, possibly some details where going wrong as I did resolve the conflicts manually.

I also already built containers for that branch, using the `develop-5206` tag.

---

However if you would like to continue working on this branch though, then I can say that yesterday evening, I updated the containers using the existing `develop-4804` tag.

**Comment by szaimen** [comment]
Hi @timdiels can you please decide on which branch we should continue to work on?

**Comment by timdiels** [comment]
@szaimen This branch just needs one more manual check on the newly built image but other than that it's ready to be merged as far as I'm concerned so it would seem less work to stick with this branch, no?

**Comment by timdiels** [comment]
On the other hand, if we then would need to repeat the work to merge it into the next major release, then I would prefer using the new branch.

**Comment by timdiels** [comment]
I manually tested my last change, it fixes the problem. Looks good to merge for me.

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [inline Containers/borgbackup/backupscript.sh:43]: Honestly, I would like to keep these config checks because what if the locla target drive is not mounted anymore? Then it will automatically create a new archive...

Same in all other places

```suggestion
if [ "$BORG_MODE" != backup ] && [ "$BORG_MODE" != test ] && ! [ -f "$BORG_BACKUP_DIRECTORY/config" ]; then
```

So you will probably need to do this check differently for local and remote...

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [inline Containers/borgbackup/backupscript.sh:158]:
```suggestion
        if ! borg info &>/dev/null; then
```

File: Containers/borgbackup/borg_excludes
Comment (szaimen) [inline Containers/borgbackup/borg_excludes:9]: we have this now in 3 places IIRC. can we maybe somehow merge this so that we only need to maintain one place?

File: php/src/Data/ConfigurationManager.php
Comment (szaimen) [inline php/src/Data/ConfigurationManager.php:467]: can we maybe somehow keep a check for local paths?

**Review by szaimen — CHANGES_REQUESTED** [review]
Sorry for coming back so late to you. 
I have some remarks. See below.
(After they are resolved, it is in a shape where we can merge this. As I said the plan is to include this in a release together with https://github.com/nextcloud/all-in-one/issues/5249)

**Comment by msullivan-lc** [comment]
Thank you for your email! I am currently out-of-office until October 17, 2024. Once I have returned I will do my best to get back to you as quickly as possible.

If this is regarding a time-sensitive matter related to Operations/Facilities please contact Stephane Sabourin  (Email: ***@***.***).

If this is regarding a time-sensitive matter related to production or any other inquiries please contact Pastor Gabe (Email: ***@***.***).

Thank you for your understanding and have a wonderful day!

**Comment by szaimen** [comment]
Hi @timdiels sorry for the ping but would you be able to look into my comments above? (I'd like to include this for the next major version of AIO which will likely come out as beta release in ~3 weeks. Until this the PR needs to be in a mergeable state. Thanks a lot already for all the work you've put into this.

File: Containers/borgbackup/borg_excludes
Comment (timdiels) [reply Containers/borgbackup/borg_excludes:9]: All 3 places have a different syntax, there's no common format, e.g. rsync only supports its own format. I would have to generate the patterns, which seems tricky in bash; it would take me a while to figure out and I think might make the code harder to read, especially for the `find`. For example:

rsync

```
        --exclude "nextcloud_aio_mastercontainer/caddy/**" \
        --exclude "nextcloud_aio_nextcloud/data/audit.log" \
```

vs find

```
                        -o -path nextcloud_aio_volumes/nextcloud_aio_mastercontainer/caddy \
                        -o -path "nextcloud_aio_volumes/nextcloud_aio_mastercontainer/caddy/*" \
                        -o -path nextcloud_aio_volumes/nextcloud_aio_nextcloud/data/audit.log \
```

vs borg

```
nextcloud_aio_volumes/nextcloud_aio_mastercontainer/caddy/
nextcloud_aio_volumes/nextcloud_aio_nextcloud/data/audit.log
```

File: php/src/Data/ConfigurationManager.php
Comment (timdiels) [reply php/src/Data/ConfigurationManager.php:467]: It's moved to `ValidateBorgLocationVars`

https://github.com/nextcloud/all-in-one/pull/4804/files/a25ebd7a60370c3f0999034fbe0d284982864535#diff-7da19c38d7098fcaf5399e1f3ee7c1d2757065e2238b8b814a81b15ef79a6d77R443

File: Containers/borgbackup/backupscript.sh
Comment (timdiels) [reply Containers/borgbackup/backupscript.sh:43]: If it's not mounted, `borg info` would fail unless there was already a valid borg repo in the unmounted directory. So I believe this works the same way as checking for the config file except that the former also works for remote.

```
❯ borg info .  # exits nonzero
/home/user/fjj is not a valid repository. Check repo config.

❯ borg info doesnotexist  # exits nonzero
Repository /home/user/fjj/doesnotexist does not exist.
```

> Then it will automatically create a new archive

I assume you mean new repo? The presence of `/nextcloud_aio_volumes/nextcloud_aio_mastercontainer/data/borg.config` will prevent initializing a new repo https://github.com/nextcloud/all-in-one/pull/4804/files/a25ebd7a60370c3f0999034fbe0d284982864535#diff-95fc189cd306db6f940ec73d973cf7b36276bff075d61dff20e17c3dcf295dabR122
which in the remote case is an empty file I create https://github.com/nextcloud/all-in-one/pull/4804/files/a25ebd7a60370c3f0999034fbe0d284982864535#diff-95fc189cd306db6f940ec73d973cf7b36276bff075d61dff20e17c3dcf295dabR168

File: Containers/borgbackup/borg_excludes
Comment (szaimen) [reply Containers/borgbackup/borg_excludes:9]: I see. Then please add a comment to all 3 places that we need to keep them in sync

**Comment by szaimen** [comment]
@timdiels one last task is then to resolve the conflicts after handling the comments. 

Afterwards I will create a final image from the branch for a last testing round.

File: Containers/borgbackup/borg_excludes
Comment (timdiels) [reply Containers/borgbackup/borg_excludes:11]: Added

**Comment by timdiels** [comment]
@szaimen I believe the only remaining comment was to keep the excludes in sync. I've merged main into it. Could you create the new image?

**Comment by szaimen** [comment]
> I believe the only remaining comment was to keep the excludes in sync. 

Yes

I've merged main into it. Could you create the new image?

I created new images 👍

File: Containers/mastercontainer/Dockerfile
Comment (szaimen) [inline Containers/mastercontainer/Dockerfile:67]: Last thing to revert before the merge

**Review by szaimen — CHANGES_REQUESTED** [review]

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [inline Containers/borgbackup/backupscript.sh:43]:
```suggestion
if [ "$BORG_MODE" != backup ] && [ "$BORG_MODE" != test ] && ! borg info &>/dev/null; then
```

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [inline Containers/borgbackup/backupscript.sh:120]:
```suggestion
    if ! borg info &>/dev/null; then
```

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [inline Containers/borgbackup/backupscript.sh:158]:
```suggestion
        if ! borg info &>/dev/null; then
```

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [inline Containers/borgbackup/backupscript.sh:531]:
```suggestion
        if ! borg info &>/dev/null; then
```

**Comment by szaimen** [comment]
@timdiels just a few suggestions to not print out borg info to not spam the logs...

File: Containers/borgbackup/backupscript.sh
Comment (timdiels) [reply Containers/borgbackup/backupscript.sh:531]: I think at least in the case, when doing the backup check, seeing the output could be useful in case it fails. It might show info about whether it's corrupt, can't connect to it, ...

Comment (timdiels) [reply Containers/borgbackup/backupscript.sh]: How about I just repeat the `borg info` in case it fails; see last commit. Then it only spams when it may be useful. There's probably a nicer way of doing this with `x=$(borg info)` and check `$?` or something instead of calling it again but this way it should work at least.

**Comment by timdiels** [comment]
I retested the remote backup, still works.

File: Containers/borgbackup/backupscript.sh
Comment (szaimen) [reply Containers/borgbackup/backupscript.sh]: `borg info >/dev/null` should do what you try to do. It only prints out if it fails.

Comment (szaimen) [reply Containers/borgbackup/backupscript.sh]: (Removing thr & from the command does this)

Comment (timdiels) [reply Containers/borgbackup/backupscript.sh]: Ah indeed, borg splits stdout/stderr nicely, I'll do that instead

**Review by szaimen — APPROVED** [review]
LGTM. Thank you very much! 🥳🎉🚀

**Comment by szaimen** [comment]
This is now released with v10.0.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Comment by KoMa1012** [comment]
Is there a way to back up to remote and local? Unfortunately I can't find any way to "reset" my old backup option (different path) somehow, even after following: [5593](https://github.com/nextcloud/all-in-one/discussions/5593#discussioncomment-11285112) (except step 5, since I'm already on 10.0.0).

Since I'm not sure how it's working currently I'll just give you how I did backups until now, once a day to local and once a week to remote. In general I never expect to use the remote only in natural disaster situations, and then I can accept loosing the last week. If this option isn't available, this could be a nice thing to add.

**Pull request merged**
